### PR TITLE
Allow destructuring of objects implementing ArrayAccess

### PR DESCRIPTION
**Pull request by herndlm**
Closes https://github.com/phpstan/phpstan/issues/6068

### CODE REVIEW

File: src/Rules/Arrays/ArrayDestructuringRule.php [hunk 1]
```diff
@@ -72,7 +73,7 @@ static function (Type $varType): bool {
 		if ($exprType instanceof ErrorType) {
 			return [];
 		}
-		if (!$exprType->isArray()->yes()) {
+		if (!$exprType->isArray()->yes() && !(new ObjectType(\ArrayAccess::class))->isSuperTypeOf($exprType)->yes()) {
```
Comment (ondrejmirtes) [inline src/Rules/Arrays/ArrayDestructuringRule.php:76]: Same thing needs to be repeated inside the `static function` a few lines above in the `findTypetoCheck` call.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

**Comment by ondrejmirtes** [comment]
Thank you!

File: src/Rules/Arrays/ArrayDestructuringRule.php (hunk 1 above)
Comment (herndlm) [reply src/Rules/Arrays/ArrayDestructuringRule.php:76]: Thx! I was still trying to figure out if and how to test that change. I was currently tinkering around with a new test class where the checkUnionTypes of the ruleLevelHelper is disabled to verify that this works. Feel free to let me know if this is needed

Comment (ondrejmirtes) [reply src/Rules/Arrays/ArrayDestructuringRule.php:76]: You can test this stuff with integration tests here: https://github.com/phpstan/phpstan-src/tree/master/tests/PHPStan/Levels

There are some test files and some JSONs. If you modify the test file and run a test where it's included (most are here): https://github.com/phpstan/phpstan-src/blob/master/tests/PHPStan/Levels/LevelsIntegrationTest.php, PHPStan will modify the JSON files to reflect the current state. The key is to realize which situations are reported on which level. In this case there are going to be different levels reported for trying to destructure `int`, `array|null`, `array|Foo`, `ArrayObject|null`, `ArrayObject|Foo` and that's where this change is important.

**Pull request merged**